pac4j-sql/src/main/java/org/pac4j/sql/credentials/authenticator/DbAuthenticator.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * @author Jerome Leleu
+ * @since 1.8.0
+ */
+public class DbAuthenticator extends AbstractUsernamePasswordAuthenticator {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    protected DBI dbi;
+
+    protected DataSource dataSource;
+
+    protected String attributes = "";
+
+    protected String startQuery = "select username, password";
+    protected String endQuery = " from users where username = :username";
+
+    public DbAuthenticator() {
+    }
+
+    public DbAuthenticator(final DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public DbAuthenticator(final DataSource dataSource, final String attributes) {
+        this.dataSource = dataSource;
+        this.attributes = attributes;
+    }
+
+    public DbAuthenticator(final DataSource dataSource, final String attributes, final PasswordEncoder passwordEncoder) {
+        this.dataSource = dataSource;
+        this.attributes = attributes;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    protected void initDbi() {
+        if (this.dbi == null) {
+            synchronized (this) {
+                if (this.dbi == null) {
+                    this.dbi = new DBI(this.dataSource);
+                }
+            }
+        }
+    }
+
+    public void validate(UsernamePasswordCredentials credentials) {
+        CommonHelper.assertNotNull("dataSource", this.dataSource);
+        CommonHelper.assertNotNull("attributes", this.attributes);
+        CommonHelper.assertNotNull("passwordEncoder", this.passwordEncoder);
+
+        initDbi();
+
+        Handle h = null;
+        try {
+            h = dbi.open();
+
+            final String username = credentials.getUsername();
+            final String query;
+            if (CommonHelper.isNotBlank(attributes)) {
+                query = startQuery + ", " + attributes + endQuery;
+            } else {
+                query = startQuery + endQuery;
+            }
+            final List<Map<String, Object>> results = h.createQuery(query).bind("username", username).list(2);
+
+            if (results == null || results.size() == 0) {
+                throw new AccountNotFoundException("No account found for: " + username);
+            } else if (results.size() > 1) {
+                throw new MultipleAccountsFoundException("Too many accounts found for: " + username);
+            } else {
+                final Map<String, Object> result = results.get(0);
+                final String expectedPassword = passwordEncoder.encode(credentials.getPassword());
+                final String returnedPassword = (String) result.get("password");
+                if (CommonHelper.areNotEquals(returnedPassword, expectedPassword)) {
+                    throw new BadCredentialsException("Bad credentials for: " + username);
+                } else {
+                    final DbProfile profile = createProfile(username, attributes.split(","), result);
```

Comment:
No, you're right: it seems pretty complicated to create a new user profile. That said, it's consistent with the entire library.
I'm not sure about the right solution, let's keep things that way and merge. We'll see that the requests will be on that topic before the release.